lib/json0.js

```diff
@@ -133,6 +133,63 @@ function convertToText(c) {
   delete c.o;
 }
 
+// not checking anything here, we should probably check that u: exists
+// (only thing we care about at json0 top level), and then delegate
+// to any subtypes if there is already subtype presence data
+json.createPresence = function(presence) {
+  return presence;
+};
+
+// this needs more thinking/testing, looking a bit more carefully at
+// how this is implemented in ot-rich-text, etc.
+json.comparePresence = function(pres1, pres2) {
+  if (!pres1 || !pres2) {
+    return false;
+  }
+  if (!pres1.p || !pres2.p) {
+    return false;
+  }
+  if (pres1.t !== pres2.t) {
+    return false;
+  }
+  if (pres1.t && subtypes[pres1.t]) {
+    if (pres1.p[0] === pres2.p[0]) {
+      return subtypes[pres1.t].comparePresence(pres1, pres2);
+    }
+  } else return pres1 === pres2;
+};
+
+// this is the key function, always run client-side, both on
+// the client that creates a text-change, and on the clients
+// that receive text-changes (ops). if there are no ops, just
+// return presence, if there are ops, delegate to the subtype
+// responsible for those ops (currently only ot-rich-text).
+// I am making assumptions many places that all ops will be
+// of the same subtype, not sure if this is a given.
+// We're only concerned about the first level of object/array,
+// not sure if the spec allows nesting of subtypes.
+json.transformPresence = function(presence, op, isOwn) {
```

Comment:
I'm thrilled to see this development!

It would be awesome if these functions were covered by tests.

Reply:
The next step would be to handle transformation of nested structures recursively.

Reply:
Also, it would be interesting to implement presence for text0, since it's embedded within json0.

Reply:
Also, it would be interesting to implement presence for text0, since it's embedded within json0.